### PR TITLE
Vendor in Chainsmoker for path matching

### PR DESCRIPTION
**Pull request by paulmelnikow**
I've rewritten [chainsmoker](https://github.com/paulmelnikow/chainsmoker) in TypeScript and vendored it in as we discussed at https://github.com/danger/danger-js/issues/431#issuecomment-480269809.

I took a stab at the integration too, though I'm not sure I've added it in all the right places. Tests are passing locally.

Looking forward to feedback!

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -93,11 +93,13 @@
     "@types/debug": "0.0.30",
     "@types/get-stdin": "^5.0.1",
     "@types/http-proxy-agent": "^2.0.1",
-    "@types/jest": "^24.0.0",
+    "@types/jest": "^24.0.11",
```
Comment (paulmelnikow) [inline package.json]: This version provides a more type-friendly version of `it.each`.

File: package.json [hunk 2]
```diff
@@ -142,8 +144,10 @@
     "lodash.includes": "^4.3.0",
     "lodash.isobject": "^3.0.2",
     "lodash.keys": "^4.0.8",
+    "lodash.mapvalues": "^4.6.0",
```
Comment (paulmelnikow) [inline package.json]: I had a hard time getting types right for my own `mapValues` implementation so used the one from lodash to save time.

**Comment by orta** [comment]
Alright, so I've given this a run through and given it all some some polish - I love it. 

I opted to take `createdAndModified` and call it `edited`. I think with the docs covering it in auto-complete, the website and the CHANGELOG covering it, it shouldn't be hard to figure out how to use it.

If you're happy, we can merge it 👍

File: package.json (hunk 2 above)
Comment (orta) [reply package.json]: Yeah, seems reasonable to me 👍

**Comment by paulmelnikow** [comment]
This looks awesome! It feels both elegant and friendly, and I hope people enjoy using it. Thanks for the docs – you described it really nicely. 👍 🚀

**Comment by orta** [comment]
Cool, lets get this shipped then

**Comment by aaronwhite** [comment]
This is great; not sure if this is possible, but would love to get the matching filePaths as part of the structure per type as well

**Comment by orta** [comment]
Maybe they could always have a `paths` attribute with that?

**Comment by aaronwhite** [comment]
So jumping right in w/ this Chainsmoker, I'm greatly simplifying the logic `if(appJSFiles.created) ...` but it's hard to report on *which* without duplicating the `fileMatch` pattern as a bunch of filters, now I have two things to maintain. I'm new to Danger, so I probably don't have all the right terms, but if it's paths (for created, modified, createdOrModified, etc) awesome!

**Comment by paulmelnikow** [comment]
> would love to get the matching filePaths as part of the structure per type as well

That makes a lot of sense.

> Maybe they could always have a `paths` attribute with that?

Yea, I think that would work. Lemme work that up!

**Comment by paulmelnikow** [comment]
Seems like this is working for us: badges/shields#3276 👍

**Pull request merged**